### PR TITLE
feat(avatar): Show presence ring and badges

### PR DESCRIPTION
**Pull request by 321gillian**
GDS-2415

- Show presence ring around avatar
- Show presence badges
- Override presence badge and display notifications badge when present

### CODE REVIEW

File: packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx [hunk 1]
```diff
@@ -71,17 +160,24 @@ export class GuxAvatar {
...
-          'gux-avatar': true,
-          [`gux-${this.size}`]: true,
-          [`gux-accent-${this.getAccent()}`]: true
-        }}
-      >
-        <slot name="image">
-          <abbr title={this.name}>{this.generateInitials()}</abbr>
-        </slot>
-      </div>
+      <Fragment>
```
Comment (jordanstith15) [inline packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx:163]: Is there a reason for this change to a wrapping fragment? I can't tell.

File: packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx [hunk 2]
```diff
@@ -61,8 +93,65 @@ export class GuxAvatar {
...
+        <div
+          class={{
+            'gux-avatar-badge': true,
+            [`${this.presence}`]: true,
+            [`gux-${this.size}`]: true
+          }}
+        >
+          <gux-icon
+            icon-name={this.getPresenceIcon(this.presence)}
+            screenreader-text={this.i18n(this.presence)}
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx:110]: I don't think we can do localisation like this. We have to allow the app to pass in the localised presence name. Maybe if they don't supply one we can fallback to this.

File: packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx (hunk 1 above)
Comment (321gillian) [reply packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx:163]: You cant have 2 children there .gux-avatar & {this.renderBadge()} with JSX. Fragment wraps them them in JSX so you don't need a div as a wrapper that doesn't do anything or has no semantic meaning.

File: packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx [hunk 3]
```diff
@@ -32,6 +40,30 @@ export class GuxAvatar {
   @Prop()
   accent: GuxAvatarAccent = 'default';
 
+  /**
+   * Shows presence such as away or available
+   */
+  @Prop()
+  presence: GuxAvatarPresence = 'available';
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx:47]: If I just use the `presence` property on its own would I be correct in saying it does create any visual change to the element. I would have to couple it with `presenceRing`

File: packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx (hunk 1 above)
Comment (jordanstith15) [reply packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx:163]: You can also return an array in that case. Fragment has special uses like Host iirc.
`
[
        <div
          class={{
            'gux-avatar': true,
            [`${this.presence}`]: true,
            [`gux-${this.size}`]: true,
            'gux-presence-ring': this.presenceRing,
            [`gux-accent-${this.getAccent()}`]: true
          }}
        >
          <div class="gux-content">
            <slot name="image">
              <abbr title={this.name}>{this.generateInitials()}</abbr>
            </slot>
          </div>
        </div>,
        {this.renderBadge()}
] as JSX.Element;
`

Comment (321gillian) [reply packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx]: I think that's a good suggestion. Applied

File: packages/genesys-spark-components/src/components/beta/gux-avatar/example.html [hunk 4]
```diff
@@ -112,13 +113,77 @@ <h2>Dynamic Example</h2>
...
+  name="EF"
+  presence="away"
+  presence-ring
+  has-badge
+></gux-avatar-beta>
+<gux-avatar-beta
+  name="MN"
+  presence="on-queue"
+  presence-ring
+  has-badge
```
[the remark this answers is not in the document]
Comment (321gillian) [reply packages/genesys-spark-components/src/components/beta/gux-avatar/example.html]: done

File: packages/genesys-spark-components/src/components/beta/gux-avatar/example.html [hunk 5]
```diff
@@ -112,13 +113,77 @@ <h2>Dynamic Example</h2>
...
+<gux-avatar-beta presence="away" name="IJ" presence-ring></gux-avatar-beta>
+<gux-avatar-beta presence="on-queue" name="KL" presence-ring></gux-avatar-beta>
+<gux-avatar-beta presence="offline" name="MN" presence-ring></gux-avatar-beta>
+<gux-avatar-beta
+  presence="out-of-office"
+  name="OP"
+  presence-ring
+></gux-avatar-beta>
+
+<h3>Presence Badge</h3>
```
[the remark this answers is not in the document]
Comment (321gillian) [reply packages/genesys-spark-components/src/components/beta/gux-avatar/example.html:242]: Added

File: packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx [hunk 6]
```diff
@@ -32,6 +40,30 @@ export class GuxAvatar {
...
+   * Shows a presence badge
+   */
+  @Prop()
+  hasBadge: boolean = false;
+
+  /**
+   * Override the presence badge with a notification icon
+   */
+  @Prop()
+  hasNotifications: boolean = false;
```
[the remark this answers is not in the document]
Comment (321gillian) [reply packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx]: Agreed

File: packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx [hunk 7]
```diff
@@ -32,6 +40,37 @@ export class GuxAvatar {
...
 
+  /**
+   * Shows presence such as away or available.
+   * Must be combined with presenceBadge or presenceRing props to take effect.
+   */
+  @Prop()
+  presence: GuxAvatarPresence = 'available';
+
+  /**
+   * Text override for presence name
```
[the remark this answers is not in the document]
Comment (321gillian) [reply packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx:51]: Updated

File: packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx [hunk 8]
```diff
@@ -32,6 +40,37 @@ export class GuxAvatar {
   @Prop()
   accent: GuxAvatarAccent = 'default';
 
+  /**
+   * Shows presence such as away or available.
+   * Must be combined with badge or ring props to take effect.
+   */
+  @Prop()
+  presence: GuxAvatarPresence = 'available';
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx:48]: This all looks good to me now except presence defaulting to "available". I think we need a "none" option that is the default.

File: packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx [hunk 9]
```diff
@@ -61,27 +100,99 @@ export class GuxAvatar {
...
+        return 'fa/subtract-circle';
+      default:
+        return 'fa/circle-xmark-regular';
+    }
+  }
+
+  private getDescriptionText(): string {
+    if (this.notifications)
+      return `${this.name} (${this.i18n('notifications')})`;
+    if (this.label) return `${this.name} (${this.label})`;
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/beta/gux-avatar/gux-avatar.tsx:160]: I also think this should be gated by `if (this.presence !== 'none')` as well as `this.label`

**Review by daragh-king-genesys — APPROVED** [review]

**Pull request merged**